workspace/resource_repo.go

```diff
@@ -0,0 +1,203 @@
+	if path, ok := r["path"]; ok {
+		err := a.client.Patch(a.context, fmt.Sprintf("/repos/%s", id), map[string]string{"path": path})
+		if err != nil {
+			return err
+		}
+		delete(r, "path")
+	}
+	return a.client.Patch(a.context, fmt.Sprintf("/repos/%s", id), r)
+}
+
+func (a ReposAPI) Read(id string) (ReposInformation, error) {
+	var resp ReposInformation
+	err := a.client.Get(a.context, fmt.Sprintf("/repos/%s", id), nil, &resp)
+	return resp, err
+}
+
+type ReposListResponse struct {
+	NextPageToken string             `json:"next_page_token,omitempty"`
+	Repos         []ReposInformation `json:"repos"`
+}
+
+func (a ReposAPI) List(prefix string) ([]ReposInformation, error) {
+	req := map[string]string{}
+	if prefix != "" {
+		req["path_prefix"] = prefix
+	}
+	reposList := []ReposInformation{}
+	for {
+		var resp ReposListResponse
+		err := a.client.Get(a.context, "/repos", req, &resp)
+		if err != nil {
+			return nil, err
+		}
+		reposList = append(reposList, resp.Repos...)
+		if resp.NextPageToken == "" {
+			break
+		}
+		req["next_page_token"] = resp.NextPageToken
+	}
+	return reposList, nil
+}
+
+func (a ReposAPI) ListAll() ([]ReposInformation, error) {
+	return a.List("")
+}
+
+var gitProvidersMap = map[string]string{
+	"github.com":    "gitHub",
+	"dev.azure.com": "azureDevOpsServices",
+	"gitlab.com":    "gitLab",
+	"bitbucket.org": "bitbucketCloud",
+}
+
+func getProviderFromUrl(uri string) string {
+	provider := ""
+	u, err := url.Parse(uri)
+	if err == nil {
+		provider = gitProvidersMap[strings.ToLower(u.Host)]
+	}
+	return provider
+}
+
+func ResourceRepo() *schema.Resource {
+	s := common.StructToSchema(ReposInformation{}, func(s map[string]*schema.Schema) map[string]*schema.Schema {
+		s["url"].ForceNew = true
+		s["url"].ValidateFunc = validation.IsURLWithScheme([]string{"https", "http"})
+		s["git_provider"].ForceNew = true
+		s["git_provider"].DiffSuppressFunc = func(k, old, new string, d *schema.ResourceData) bool {
+			return strings.EqualFold(old, new)
+		}
+		s["path"].ForceNew = true // TODO: remove after the Update API will support changing the path
+		s["branch"].ConflictsWith = []string{"tag"}
+		s["branch"].ValidateFunc = validation.StringIsNotWhiteSpace
+
+		s["tag"] = &schema.Schema{
+			Type:          schema.TypeString,
+			Optional:      true,
+			ConflictsWith: []string{"branch"},
+			ValidateFunc:  validation.StringIsNotWhiteSpace,
+		}
+
+		delete(s, "id")
+		return s
+	})
+
+	return common.Resource{
+		Schema:        s,
+		SchemaVersion: 1,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			reposAPI := NewReposAPI(ctx, c)
+			req := createRequest{Path: d.Get("path").(string), Provider: d.Get("git_provider").(string), Url: d.Get("url").(string)}
+			resp, err := reposAPI.Create(req)
+			if err != nil {
+				return err
+			}
+			d.SetId(resp.RepoID())
+			branch := d.Get("branch").(string)
```

Comment:
technically, you can re-use the update method here, right?..

Reply:
update logic is different here compared to Update method - I don't want to re-checkout branch if it's already checked out. Correct way would be if the Create API had the `branch` and `tag` parameters, but it's not implemented there - that's why two calls.  On other side, I can specify `branch` parameter set to the default branch of repository, that's why we have `branch != resp.Branch` condition